File: hcemi-appdev-pims/MI.PIMS.MT/MI.PIMS.BL/Common/Telemetry/LogInfo.cs
#nullable enable

using Microsoft.AspNetCore.Http;
using System.Collections.Generic;
using System.Diagnostics;
using System;
using System.Linq;
using System.Text.Json;

namespace MI.PIMS.BL.Common.Telemetry
{
    public class LogInfo
    {
        public string Operation { get; set; } = string.Empty;
        public string Status { get; set; } = "Success";
        public string Source { get; set; } = "PIMS UI";
        public string? UserId { get; set; }
        public string? CorrelationId { get; set; }
        public object? AdditionalInfo { get; set; }
        public DateTime Timestamp { get; set; } = DateTime.UtcNow;
        public string Type { get; set; } = string.Empty;

        public void AutoFillCorrelationId(HttpContext? context)
        {
            if (string.IsNullOrEmpty(CorrelationId))
            {
                CorrelationId = context?.Request.Headers["X-Correlation-ID"].FirstOrDefault()
                                ?? Activity.Current?.TraceId.ToString()
                                ?? context?.TraceIdentifier;
            }
        }

        public string GetAdditionalInfo()
        {
            if (AdditionalInfo == null)
            {
                return string.Empty;
            }

            if (AdditionalInfo is string)
            {
                return AdditionalInfo.ToStringNullSafe();
            }
            else
            {
                try
                {
                    return JsonSerializer.Serialize(AdditionalInfo, new JsonSerializerOptions { WriteIndented = true });
                }
                catch(Exception)
                {
                    return AdditionalInfo.ToStringOrNull();
                }
            }
        }

        public Dictionary<string, string> ToDictionary()
        {
            var dict = new Dictionary<string, string>
            {
                { "Operation", Operation },
                { "Status", Status },
                { "Source", Source },
                { "Timestamp", Timestamp.ToString("o") }
            };

            if (!string.IsNullOrEmpty(UserId)) dict["UserId"] = UserId;
            if (!string.IsNullOrEmpty(CorrelationId)) dict["CorrelationId"] = CorrelationId;
            if (AdditionalInfo != null) dict["AdditionalInfo"] = GetAdditionalInfo();
            if (!string.IsNullOrEmpty(Type)) dict["Type"] = Type;

            return dict;
        }
    }
}

File: hcemi-appdev-pims/MI.PIMS.UI/MI.PIMS.UI/wwwroot/js/Telemetry/LogInfo.js
class LogInfo {
    constructor(operationName, userId, additionalInfo, type = "", correlationId = "", status = "Success", source = "PIMS UI", timestamp = new Date()) {
        this.operationName = operationName;
        this.userId = userId;
        this.status = status;
        this.source = source;
        this.correlationId = correlationId;
        this.type = type;
        this.additionalInfo = additionalInfo;
        this.timestamp = timestamp.toISOString();
    };

    getAdditionalInfo() {
        try {
            const parsed = JSON.parse(this.additionalInfo);

            if (typeof parsed === 'object' && parsed !== null) {
                return JSON.stringify(this.additionalInfo);
            }
            else {
                return this.additionalInfo.toString();
            }
        }
        catch {
            return this.additionalInfo.toString();
        }
    };

    // Method to serialize the object properties to a format suitable for Application Insights
    toTelemetryProperties() {
        return {
            "OperationName": this.operationName,
            "UserId": this.userId,
            "Status": this.status,
            "Source": this.source,
            "CorrelationId": this.correlationId,
            "Type": this.type,
            "Timestamp": this.timestamp,
            "AdditionalInfo": this.getAdditionalInfo()
        };
    };
};

